### DISCUSSION TITLE
Issue with DHCP plugin and classless static routing.

### DISCUSSION THREAD

**Discussion by Molgaard78** [strongswan/strongswan#1055] [Q&A] [answered]
I am trying to create a split tunneling solution using Strongswan and have runned into a bit of an issue.

The setup is basically this:

Client (192.168.201.9) <--> StrongSwan VPN server  (192.168.200.1) <--> DHCP server (192.168.200.4).

The idea is that the DHCP server assigns the ip adress to the client and send it DHCP option 121 (classless static routing) or DHCP option 249 (Microsoft classless static routing) along the other options such as DHCP option 6 (DNS).

However: 

I can see through debugging the DHCP server in Wireshark, that the DHCP server does indeed send option 121 and option 249:

![Output from DHCP server](https://user-images.githubusercontent.com/96376712/169267346-a69e04ec-ba49-4ae3-863b-637d95c817d4.jpg)

But I cannot see that this information is passed onto the client?

![Output from client](https://user-images.githubusercontent.com/96376712/169267959-e7aadd74-ecbf-40f8-8c64-dd49c8335bb8.jpg)

I know it looks a bit strange that the DHCP server tells the client that the client itself is standard gateway for the subnets 10.88.0.0 and 10.0.200.0, that is the only way I could get Windows to accept a static route over VPN, when I entered it manually on client. Windows rejected the route when setting standard gateway to 0.0.0.0 as suggested by https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_split_tunneling_with_ikev2

I thought the purpose of Strongswan dhcp plugin was to act as a proxy between client and DHCP server and pass information back and forth between client and dhcp server a verbatim?

**Comment by tobiasbrunner** [comment, marked as answer]
> I thought the purpose of Strongswan dhcp plugin was to act as a proxy between client and DHCP server and pass information back and forth between client and dhcp server a verbatim?

But only IP address and DNS servers, which it forwards via IKE. Did you read [the docs](https://docs.strongswan.org/docs/5.9/plugins/dhcp.html)?

**Reply by Molgaard78** [reply]
I did back and forth.

I know that Strongswan can block for traffic that is intended for remote network by setting traffic scope to contain only the subnets I want access to through the VPN.

However: What you end up with is that windows clients can either access the ressources available on the VPN or being able to communicate with the rest of the Internet.

But you cannot selected traffic from the client to VPN ressources and remaining traffic to the Internet at the same time, due to Windows clients set the default gateway to be the VPN connection.

I have intentionally blocked that VPN clients can access the Internet through the VPN, hence the either/or scenario.

If I set the client to not use the VPN server as standard gateway, then I cannot access any subnet available through the VPN due to client routing table does not know which connection to use to access the subnets.

This is why I want to pass the classless static routing option from the DHCP server onto the client, because that will update the routing table and get the end result I want:

That is only the wanted subnets is available through the VPN and everything else goes to the Internet through standard gateway for whatever LAN the machine is on.

**Reply by tobiasbrunner** [reply]
> But you cannot selected traffic from the client to VPN ressources and remaining traffic to the Internet at the same time, due to Windows clients set the default gateway to be the VPN connection.

You have to disable that option (or enable the `SplitTunneling` option via PowerShell) and then install routes manually via [Add-VpnConnectionRoute](https://docs.microsoft.com/en-us/powershell/module/vpnclient//Add-VpnConnectionRoute?view=win10-ps), if the virtual IP is not from the tunneled remote subnet (if there are multiple, you probably can't avoid having to add routes manually).

If sending DHCP via IPsec tunnel is an option, I don't know. I think it might have been at some point with Windows 10 Mobile. But no idea if or how that ever worked with regular Windows clients. But it definitely doesn't involve the _dhcp_ plugin.

**Reply by Molgaard78** [reply]
... in other words: Make it as difficult to manage a network with split tunneling as possible, since using Powershell to solve everything is litterally **not an option**.

Particularly since the DHCP protocol in itself has a solution for the problem. Namely through option 121 and 249.

The powershell solution requires that it gets triggered both when VPN goes up, so routing to subnets can occur and another powershell is triggered on teardown of VPN or else they might run in unintentionally side effects due to client routing table contains rules that should not be there if VPN is not up.

If I want to automate the scripts beyond this point then I basically need that all clients are connected to the network as domain clients, which is a particular non-starter for me, due to all clients are using their private device.

- 

My main goal with Strongswan was to create a VPN solution with split tunneling that was as user friendly as possible for end users.

You really can't get it more user friendly than telling them: Here is a username and password and tell them which server to use and the press connect.

The alternative is of course WireGuard, which makes administration of client keys a nightmare and OpenVPN that requires specially designed certificates and beyond this also suffer performance wise vs Strongswan and WireGuard.

The ideal solution with Strongswan would be that the line (server side):

local-ts = 10.88.0.0/16, 10.0.200.0/24

Would by itself tell the client to only route those two subnet over VPN.

In that case I do not need an external DHCP, but just define the pool directly in swanctl.conf.

**Reply by tobiasbrunner** [reply]
> The powershell solution requires that it gets triggered both when VPN goes up, so routing to subnets can occur and another powershell is triggered on teardown of VPN or else they might run in unintentionally side effects due to client routing table contains rules that should not be there if VPN is not up.

Not at all. The configured routes are persistently associated with the VPN connection and get automatically installed/removed with it. You could write a script that creates the VPN connection, with stronger algorithms (which can't be done to the full extent without PowerShell anyway) and added routes etc. Then the connection can be initiated via GUI without further scripting.

> The ideal solution with Strongswan would be that the line (server side):

Yep, that's exactly how it works with some clients (that support split-tunneling via narrowing). And with Windows it works too if there is only a single subnet and the virtual IPs are assigned from it (for which you could use the _dhcp_ and _farp_ plugins). But if you want to do split-tunneling to multiple subnets you have to manually configure the routes.

**Reply by Molgaard78** [reply]
Okay.

I will dig more into the powershell solution as it is the lesser evil for my needs.

On the plus side it simplifies things on the server side.
